crates/core/src/sql/execute.rs

```diff
@@ -79,26 +75,38 @@ pub fn execute_single_sql(
 #[tracing::instrument(skip_all)]
 pub fn execute_sql(
     db: &RelationalDB,
-    tx: &mut MutTx,
     ast: Vec<CrudExpr>,
     query_debug_info: Option<&QueryDebugInfo>,
     auth: AuthCtx,
 ) -> Result<Vec<MemTable>, DBError> {
     let total = ast.len();
     let ctx = ExecutionContext::sql(db.address(), query_debug_info);
-    let p = &mut DbProgram::new(&ctx, db, tx, auth);
-    let q = Expr::Block(ast.into_iter().map(|x| Expr::Crud(Box::new(x))).collect());
-
     let mut result = Vec::with_capacity(total);
-    collect_result(&mut result, run_ast(p, q).into())?;
+    match CrudExpr::is_reads(&ast) {
+        false => db.with_auto_commit(&ctx, |mut_tx| {
+            let mut tx: TxMode = mut_tx.into();
+            let q = Expr::Block(ast.into_iter().map(|x| Expr::Crud(Box::new(x))).collect());
+            let p = &mut DbProgram::new(&ctx, db, &mut tx, auth);
+            collect_result(&mut result, run_ast(p, q).into())
+        }),
+        true => db.with_read_only(&ctx, |tx| {
+            let mut tx: TxMode = tx.into();
+            let q = Expr::Block(ast.into_iter().map(|x| Expr::Crud(Box::new(x))).collect());
+            let p = &mut DbProgram::new(&ctx, db, &mut tx, auth);
+            collect_result(&mut result, run_ast(p, q).into())
+        }),
+    }?;
+
     Ok(result)
 }
 
 /// Run the `SQL` string using the `auth` credentials
 #[tracing::instrument(skip_all)]
-pub fn run(db: &RelationalDB, tx: &mut MutTx, sql_text: &str, auth: AuthCtx) -> Result<Vec<MemTable>, DBError> {
-    let ast = compile_sql(db, tx, sql_text)?;
-    execute_sql(db, tx, ast, Some(&QueryDebugInfo::from_source(sql_text)), auth)
+pub fn run(db: &RelationalDB, sql_text: &str, auth: AuthCtx) -> Result<Vec<MemTable>, DBError> {
+    let query_info = &QueryDebugInfo::from_source(&sql_text);
+    let ctx = &ExecutionContext::sql(db.address(), Some(query_info));
+    let ast = db.with_read_only(&ctx, |tx| compile_sql(db, tx, sql_text))?;
+    execute_sql(db, ast, Some(&QueryDebugInfo::from_source(sql_text)), auth)
```

Comment:
Are compilation and execution in two separate transactions now?

Reply:
yes, compilation works with read-only tx, then execution could be read-tx or write-tx based on compilation result (ast).

Reply:
I think that's fine. Can you just add a comment to that effect since it might be unexpected?